src/oidc-challenge.ts

```diff
@@ -28,65 +41,92 @@ const invalidAccessCode: ServerChallengeResponse = {
+export type AssignmentPolicy = (
+  subjectName: Name,
+  userId: string,
+) => Promise<Name | undefined>;
 
+export class ServerOidcChallenge implements ServerChallenge<State> {
   constructor(
-      private readonly requestHeader: Record<string, string>,
-      private requestBody: URLSearchParams,
-      private readonly requestUrl: string,
-      private readonly pubKeyUrl: string,
-      private readonly assignmentPolicy?: ServerOidcChallenge.AssignmentPolicy,
+    public readonly challengeId: string,
+    public readonly timeLimit: number,
+    public readonly retryLimit: number,
+    private readonly options: {
+      requestHeader: Record<string, string>;
+      requestBody: URLSearchParams;
+      requestUrl: string;
+      pubKeyUrl: string;
+      assignmentPolicy?: AssignmentPolicy;
+    },
   ) {}
 
-  public async process(request: ChallengeRequest, context: ServerChallengeContext<State>): Promise<ServerChallengeResponse> {
-    
+  public async process(
+    request: ChallengeRequest,
+    context: ServerChallengeContext<State>,
+  ): Promise<ServerChallengeResponse> {
     const {
       "oidc-id": oidcId,
       "access-code": accessCode,
     } = request.parameters;
+    console.info(
+      `Challenge request: ${
+        JSON.stringify({
+          oidcId: fromUtf8(oidcId),
+          accessCode: fromUtf8(accessCode),
+        })
+      }`,
+    );
     if (!oidcId || !accessCode) {
       return invalidParameters;
     }
     context.challengeState = { oidcId, accessCode };
     // write access code to the request body
-    this.requestBody.append("code", fromUtf8(accessCode));
+    this.options.requestBody.append("code", fromUtf8(accessCode));
     try {
-      const response = await fetch(this.requestUrl, {
-        method: 'post',
-        body: this.requestBody,
-        headers: this.requestHeader
+      const response = await fetch(this.options.requestUrl, {
+        method: "post",
+        body: this.options.requestBody,
+        headers: this.options.requestHeader,
       });
       const data = await response.json();
-      const JWKS = createRemoteJWKSet(new URL(this.pubKeyUrl));
-      console.log(data)
+      const JWKS = createRemoteJWKSet(new URL(this.options.pubKeyUrl));
       if (data["error"]) {
+        console.error(`Invalid AccessToken: ${JSON.stringify(data)}`);
         return invalidAccessCode;
-      }
-      else if (data["id_token"]) {
+      } else if (data["id_token"]) {
         const { payload } = await jwtVerify(data["id_token"], JWKS);
-        console.log(payload)
-        try { 
-          await this.assignmentPolicy?.(context.subjectName, String(payload["email"]));
+        try {
+          const assignedName = await this.options.assignmentPolicy?.(
+            context.subjectName,
+            String(payload["email"]),
+          );
+          if (assignedName) {
+            // Force the certificate to be renamed
+            const contextInternal = context as unknown as {
```

Comment:
1/ I will merge this anyway
2/ Just for discussion and justification: this is too hacky. `certRequestPub` is designed to be read-only, which implies the NDNts CA is supposed to issue the cert the requester applied -- that was why I disagreed and still disagree with forced renaming.

Reply:
We can ask Junxiao to change his code, and we can change correspondingly. NDNts locks too many fields, and it is necessary to break the visibility sometimes. Just do it.

`certRequestPub` is only a tuple of a Name and the public key. It is not a certificate and there is no need to maintain any relation among them.